### PR TITLE
Add support to cast decimal to double

### PR DESCRIPTION
**Pull request by karteekmurthys**
Added support for converting long and short decimals to double. Includes unit tests.

### CODE REVIEW

**Review by majetideepak — COMMENTED** [review]
Add a test for null input as well.

**Comment by karteekmurthys** [comment]
@Yuhta would you please review this.

**Comment by majetideepak** [comment]
@karteekmurthys can we hold on to merging this until https://github.com/facebookincubator/velox/pull/4434?

**Review by Yuhta — CHANGES_REQUESTED** [review]
Need rebase onto https://github.com/facebookincubator/velox/pull/4434

**Review by majetideepak — APPROVED** [review]

**Comment by karteekmurthys** [comment]
@Yuhta would you please review this change.

**Comment by karteekmurthys** [comment]
@Yuhta would you please review this PR.

**Comment by karteekmurthys** [comment]
@Yuhta @mbasmanova would you please review this.

File: velox/expression/CastExpr.cpp [hunk 1]
```diff
@@ -59,6 +59,17 @@ void applyCastKernel(
...
     writer.finalize();
+  } else if (isDecimalType(*input->type())) {
+    // Special handling for decimal types.
+    auto [precision, scale] = getDecimalPrecisionScale(*input->type());
+    if constexpr (ToKind == TypeKind::DOUBLE) {
+      auto output =
+          util::Converter<CppToType<double>::typeKind, void, Truncate>::cast(
+              input->valueAt(row), nullOutput);
+      if (!nullOutput) {
+        result->set(row, output / (double)DecimalUtil::kPowersOfTen[scale]);
```
Comment (jinchengchenghh) [inline velox/expression/CastExpr.cpp:70]: Does this satisfy the divide half round up?

Comment (Yuhta) [reply velox/expression/CastExpr.cpp:70]: I don't think rounding is an issue here

**Comment by Yuhta** [comment]
@karteekmurthys Can you resolve the conflict?

**Comment by Yuhta** [comment]
@karteekmurthys The build is still not passing, probably some API change in Type interface

**Comment by karteekmurthys** [comment]
> @karteekmurthys The build is still not passing, probably some API change in Type interface

Yes, looking into it. There have been some rewrites in this area. Will resolve them

**Comment by Yuhta** [comment]
@karteekmurthys Can you check if the regression on benchmark is related?

File: velox/expression/CastExpr.cpp [hunk 2]
```diff
@@ -46,6 +47,17 @@ void applyCastKernel(
     vector_size_t row,
     const SimpleVector<typename TypeTraits<FromKind>::NativeType>* input,
     FlatVector<typename TypeTraits<ToKind>::NativeType>* result) {
+  if (input->type()->isDecimal()) {
```
Comment (Yuhta) [inline velox/expression/CastExpr.cpp:50]: Do we need a compile time check on input kind?

Comment (karteekmurthys) [reply velox/expression/CastExpr.cpp:50]: You mean why is it a runtime check? Decimal Types are logical types and there are no TypeKinds mapped to them. So we can't make it a compile time check. The mean runtime is 10183928.214538805 ns vs 101502160.15259695 ns. There is a difference of 0.3ms. CMIW.

I do see some cast operations in the benchmarking that could hit this check and result in slowdown:

```
BENCHMARK(multiplyLiteralSmall) {
  benchmark->runSmall("multiply(a, cast(2 as double))");
}

BENCHMARK(multiplyLiteralMedium) {
  benchmark->runMedium("multiply(a, cast(2 as double))");
}

BENCHMARK(multiplyLiteralLarge) {
  benchmark->runLarge("multiply(a, cast(2 as double))");
}
```

Comment (karteekmurthys) [reply velox/expression/CastExpr.cpp:50]: I will try to isolate applyCastKernel for Decimal case into a separate method may be, so that the runtime check is not made for every row.
